goplay/src/main/java/be/tapped/goplay/stream/StreamRepo.kt

```diff
@@ -0,0 +1,127 @@
+package be.tapped.goplay.stream
+
+import arrow.core.Either
+import arrow.core.Either.Companion.catch
+import arrow.core.computations.either
+import arrow.core.left
+import be.tapped.goplay.ApiResponse
+import be.tapped.goplay.apiGoPlay
+import be.tapped.goplay.apiVierVijfZes
+import be.tapped.goplay.content.Program
+import be.tapped.goplay.profile.IdToken
+import be.tapped.goplay.safeGet
+import io.ktor.client.HttpClient
+import io.ktor.client.request.headers
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.jsonObject
+import kotlinx.serialization.json.jsonPrimitive
+import org.apache.http.HttpHeaders
+
+internal fun interface StreamRepo {
+    suspend fun streamByVideoUuid(videoId: Program.Detail.Playlist.Episode.VideoUuid, idToken: IdToken): Either<ApiResponse.Failure, ApiResponse.Success.Stream>
+}
+
+internal fun httpStreamRepo(client: HttpClient, mpegDashStreamResolver: MpegDashStreamResolver, hlsStreamResolver: HLSStreamResolver): StreamRepo =
+    StreamRepo { videoId, idToken ->
+        either {
+            val videoResponse = client.safeGet<JsonElement>("$apiVierVijfZes/content/${videoId.videoUuid}") { headers { append(HttpHeaders.AUTHORIZATION, idToken.token) } }.bind()
+            if (videoResponse is JsonObject) {
+                val videoObj = videoResponse.jsonObject
+                when {
+                    videoObj.containsKey("videoDash") -> mpegDashStreamResolver.fetchMpegDashStream(videoId, videoObj, idToken)
+                    videoObj.containsKey("video") -> hlsStreamResolver.fetchHlsStream(videoId, videoObj)
+                    else -> ApiResponse.Failure.Stream.UnknownStream(videoId, videoObj).left()
+                }.bind()
+            } else {
+                ApiResponse.Success.Stream(ResolvedStream.NoStreamFound(videoId))
+            }
+        }
+    }
+
+//<editor-fold desc="MPEG-DASH">
+internal fun interface MpegDashStreamResolver {
+    suspend fun fetchMpegDashStream(
+        videoId: Program.Detail.Playlist.Episode.VideoUuid,
+        videoObj: JsonObject,
+        idToken: IdToken
+    ): Either<ApiResponse.Failure, ApiResponse.Success.Stream>
+}
+
+/**
+ * Example of MPEG dash response
+ *
+ * ```json
+ * {
+ *   "drmKey": {
+ *     "S": "very-secret-drm-key"
+ *   },
+ *   "videoDash": {
+ *     "S": "https://www.mpg-dash.be"
+ *   },
+ *   "video": {
+ *     "S": "https://www.hls-video.be"
+ *   }
+ * }
+ * ```
+ *
+ * Example of DRM response
+ *
+ * ```json
+ * {
+ *   "auth": "secret-authentication-token"
+ * }
+ * ```
+*/
+internal fun mpegDashStreamResolver(client: HttpClient): MpegDashStreamResolver =
```

Comment:
Maybe the MpegDashStreamResolver should piggy back on the HlsStreamResolver if it fails.